esphome/components/dps310/dps310.cpp

```diff
@@ -0,0 +1,193 @@
+    this->mark_failed();
+    return;
+  }
+  delay(10);
+  // wait for the sensor and its coefficients to be ready
+  while (timer-- && (!(reg & DPS310_BIT_SENSOR_RDY) || !(reg & DPS310_BIT_COEF_RDY))) {
+    reg = this->read_byte(DPS310_REG_MEAS_CFG).value_or(0);
+    delay(5);
+  }
+
+  if (!(reg & DPS310_BIT_SENSOR_RDY) || !(reg & DPS310_BIT_COEF_RDY)) {  // the flags were not set in time
+    this->mark_failed();
+    return;
+  }
+  // read device ID
+  if (!this->read_byte(DPS310_REG_PROD_REV_ID, &this->prod_rev_id_)) {
+    this->mark_failed();
+    return;
+  }
+  // read in coefficients used to calculate the compensated pressure and temperature values
+  if (!this->read_bytes(DPS310_REG_COEF, coef_data_raw, DPS310_NUM_COEF_REGS)) {
+    this->mark_failed();
+    return;
+  }
+  // read in coefficients source register, too -- we need this a few lines down
+  if (!this->read_byte(DPS310_REG_TMP_COEF_SRC, &reg)) {
+    this->mark_failed();
+    return;
+  }
+  // set up operational stuff
+  if (!this->write_byte(DPS310_REG_PRS_CFG, DPS310_VAL_PRS_CFG)) {
+    this->mark_failed();
+    return;
+  }
+  if (!this->write_byte(DPS310_REG_TMP_CFG, DPS310_VAL_TMP_CFG | (reg & DPS310_BIT_TMP_COEF_SRC))) {
+    this->mark_failed();
+    return;
+  }
+  if (!this->write_byte(DPS310_REG_CFG, DPS310_VAL_REG_CFG)) {
+    this->mark_failed();
+    return;
+  }
+  if (!this->write_byte(DPS310_REG_MEAS_CFG, 0x07)) {  // enable background mode
+    this->mark_failed();
+    return;
+  }
+
+  this->c0_ =  // we only ever use c0/2, so just divide by 2 here to save time later
+      DPS310Component::twos_complement(
+          int16_t(((uint16_t) coef_data_raw[0] << 4) | (((uint16_t) coef_data_raw[1] >> 4) & 0x0F)), 12) /
+      2;
+
+  this->c1_ =
+      DPS310Component::twos_complement(int16_t((((uint16_t) coef_data_raw[1] & 0x0F) << 8) | coef_data_raw[2]), 12);
+
+  this->c00_ = ((uint32_t) coef_data_raw[3] << 12) | ((uint32_t) coef_data_raw[4] << 4) |
+               (((uint32_t) coef_data_raw[5] >> 4) & 0x0F);
+  this->c00_ = DPS310Component::twos_complement(c00_, 20);
+
+  this->c10_ =
+      (((uint32_t) coef_data_raw[5] & 0x0F) << 16) | ((uint32_t) coef_data_raw[6] << 8) | (uint32_t) coef_data_raw[7];
+  this->c10_ = DPS310Component::twos_complement(c10_, 20);
+
+  this->c01_ =
+      DPS310Component::twos_complement(int16_t(((uint16_t) coef_data_raw[8] << 8) | (uint16_t) coef_data_raw[9]), 16);
+  this->c11_ =
+      DPS310Component::twos_complement(int16_t(((uint16_t) coef_data_raw[10] << 8) | (uint16_t) coef_data_raw[11]), 16);
+  this->c20_ =
+      DPS310Component::twos_complement(int16_t(((uint16_t) coef_data_raw[12] << 8) | (uint16_t) coef_data_raw[13]), 16);
+  this->c21_ =
+      DPS310Component::twos_complement(int16_t(((uint16_t) coef_data_raw[14] << 8) | (uint16_t) coef_data_raw[15]), 16);
+  this->c30_ =
+      DPS310Component::twos_complement(int16_t(((uint16_t) coef_data_raw[16] << 8) | (uint16_t) coef_data_raw[17]), 16);
```

Comment:
These are full 16-bit numbers, so you wouldn't need to do the twos-compliment math.
```suggestion
  this->c01_ = (int16_t)((uint16_t) coef_data_raw[8] << 8) | (uint16_t) coef_data_raw[9]);
  this->c11_ = (int16_t)((uint16_t) coef_data_raw[10] << 8) | (uint16_t) coef_data_raw[11]);
  this->c20_ = (int16_t)((uint16_t) coef_data_raw[12] << 8) | (uint16_t) coef_data_raw[13]);
  this->c21_ = (int16_t)((uint16_t) coef_data_raw[14] << 8) | (uint16_t) coef_data_raw[15]);
  this->c30_ = (int16_t)((uint16_t) coef_data_raw[16] << 8) | (uint16_t) coef_data_raw[17]);
```

Reply:
Great catch! It's interesting, everybody else does it anyway (and admittedly I largely copy/pasted this and didn't think much about it). Anything we could be missing? 😄 References: [Adafruit](https://github.com/adafruit/Adafruit_DPS310/blob/fe563adb0389aa630ffa44b1fff3482212711a11/Adafruit_DPS310.cpp#L210-L214), [Infineon](https://github.com/Infineon/DPS310-Pressure-Sensor/blob/ed02f803fc780cbcab54ed8b35dd3d718f2ebbda/src/Dps310.cpp#L110-L120)